misc/python/materialize/cli/mzcompose.py

```diff
@@ -78,6 +79,25 @@ def main(argv: List[str]) -> int:
                 print(f"    {path.relative_to(Path.cwd())}", file=sys.stderr)
         return 1
 
+    # Load the mzworkflows.py file, if any
+    mzworkflows_py = os.path.join(args.mz_find or Path.cwd(), "mzworkflows.py")
+
+    if os.path.isfile(mzworkflows_py):
+        module_name = PurePath(mzworkflows_py).parent.name
+
+        # Those lines are lifted verbatim from the importlib doc, so silence mypy warnings
+        spec = importlib.util.spec_from_file_location(module_name, mzworkflows_py)
+        module = importlib.util.module_from_spec(spec)  # type: ignore
+        spec.loader.exec_module(module)  # type: ignore
+
+        python_funcs = list(
+            filter(
+                lambda x: x[0].startswith("workflow"), getmembers(module, isfunction)
+            )
+        )
+        for python_func in python_funcs:
+            composition.python_funcs[python_func[0]] = python_func[1]
```

Comment:
So this stanza looks good, but I think we should consider it part of the composition—after all, it results in modifying `composition.python_funcs`—and put it in `Composition.__init__`.

Reply:
So, this is where I'd recommend putting it, plus a few `assert`s to keep mypy happy without having to resort to the `# type: ignore` hammer:

```diff
diff --git a/bin/pycheck b/bin/pycheck
index c8265223b..d9827b04a 100755
--- a/bin/pycheck
+++ b/bin/pycheck
@@ -31,6 +31,7 @@ try xargs bin/pyactivate --dev -m mypy \
     --no-error-summary \
     --namespace-packages \
     --explicit-package-bases \
+    --python-version 3.8 \
     <<< "$py_files"
 try bin/pyactivate --dev -m pytest -qq --doctest-modules misc/python
 
diff --git a/misc/python/materialize/cli/mzcompose.py b/misc/python/materialize/cli/mzcompose.py
index 874231acd..d6cb4f63a 100644
--- a/misc/python/materialize/cli/mzcompose.py
+++ b/misc/python/materialize/cli/mzcompose.py
@@ -10,14 +10,12 @@
 # mzcompose.py ‚Äî runs Docker Compose with Materialize customizations.
 
 import argparse
-import importlib.util
 import json
 import os
 import subprocess
 import sys
 import webbrowser
-from inspect import getmembers, isfunction
-from pathlib import Path, PurePath
+from pathlib import Path
 from typing import IO, List, Optional, Sequence, Text, Tuple
 
 from materialize import errors, mzbuild, mzcompose, spawn, ui
@@ -79,25 +77,6 @@ def main(argv: List[str]) -> int:
                 print(f"    {path.relative_to(Path.cwd())}", file=sys.stderr)
         return 1
 
-    # Load the mzworkflows.py file, if any
-    mzworkflows_py = os.path.join(args.mz_find or Path.cwd(), "mzworkflows.py")
-
-    if os.path.isfile(mzworkflows_py):
-        module_name = PurePath(mzworkflows_py).parent.name
-
-        # Those lines are lifted verbatim from the importlib doc, so silence mypy warnings
-        spec = importlib.util.spec_from_file_location(module_name, mzworkflows_py)
-        module = importlib.util.module_from_spec(spec)  # type: ignore
-        spec.loader.exec_module(module)  # type: ignore
-
-        python_funcs = list(
-            filter(
-                lambda x: x[0].startswith("workflow"), getmembers(module, isfunction)
-            )
-        )
-        for python_func in python_funcs:
-            composition.python_funcs[python_func[0]] = python_func[1]
-
     # Handle special mzcompose commands that apply to the composition.
     if args.command == "list-workflows":
         return list_workflows(composition)
diff --git a/misc/python/materialize/mzcompose.py b/misc/python/materialize/mzcompose.py
index 000d93203..c0907a7dd 100644
--- a/misc/python/materialize/mzcompose.py
+++ b/misc/python/materialize/mzcompose.py
@@ -20,6 +20,9 @@ import ipaddress
 import json
 import os
 import pathlib
+from inspect import getmembers, isfunction
+import importlib
+import importlib.util
 import random
 import re
 import shlex
@@ -179,7 +182,7 @@ class Composition:
         self.name = name
         self.repo = repo
         self.images: List[mzbuild.Image] = []
-        self.python_funcs: Dict[str, Callable] = {}
+        self.python_funcs: Dict[str, Callable[[Composition], None]] = {}
 
         default_tag = os.getenv(f"MZBUILD_TAG", None)
 
@@ -194,6 +197,18 @@ class Composition:
         # Stash away sub workflows so that we can load them with the correct environment variables
         self.yaml_workflows = compose.pop("mzworkflows", None)
 
+        # Load the mzworkflows.py file, if any.
+        mzworkflows_py = self.path.parent / "mzworkflows.py"
+        if mzworkflows_py.exists():
+            spec = importlib.util.spec_from_file_location("mzworkflows", mzworkflows_py)
+            assert spec
+            module = importlib.util.module_from_spec(spec)
+            assert isinstance(spec.loader, importlib.abc.Loader)
+            spec.loader.exec_module(module)
+            for name, fn in getmembers(module, isfunction):
+                if name.startswith("workflow_"):
+                    self.python_funcs[name] = fn
+
         # Resolve all services that reference an `mzbuild` image to a specific
         # `image` reference.
         for name, config in compose["services"].items():
```